File: src/metadata/xmp.ts
import type { PdfRef } from '../core/types.js';
import { ObjectStore } from '../core/object-store.js';
import { pdfStream, pdfName } from '../core/objects.js';
import type { DocumentInfo } from './info-dict.js';

function escapeXml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function toISODate(d: Date): string {
  return d.toISOString();
}

/**
 * Generate an XMP metadata XML packet.
 * Supports Dublin Core, XMP basic, PDF properties, and optional PDF/A identification.
 */
export function createXMPMetadata(info: DocumentInfo, pdfALevel?: 'a' | 'b' | 'u'): string {
  const parts: string[] = [];

  parts.push('<?xpacket begin="\uFEFF" id="W5M0MpCehiHzreSzNTczkc9d"?>');
  parts.push('<x:xmpmeta xmlns:x="adobe:ns:meta/">');
  parts.push('<rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">');
  parts.push('<rdf:Description rdf:about=""');
  parts.push('  xmlns:dc="http://purl.org/dc/elements/1.1/"');
  parts.push('  xmlns:xmp="http://ns.adobe.com/xap/1.0/"');
  parts.push('  xmlns:pdf="http://ns.adobe.com/pdf/1.3/"');
  if (pdfALevel) {
    parts.push('  xmlns:pdfaid="http://www.aiim.org/pdfa/ns/id/"');
  }
  parts.push('>');

  // Dublin Core
  if (info.title) {
    parts.push('<dc:title>');
    parts.push('  <rdf:Alt>');
    parts.push(`    <rdf:li xml:lang="x-default">${escapeXml(info.title)}</rdf:li>`);
    parts.push('  </rdf:Alt>');
    parts.push('</dc:title>');
  }

  if (info.author) {
    parts.push('<dc:creator>');
    parts.push('  <rdf:Seq>');
    parts.push(`    <rdf:li>${escapeXml(info.author)}</rdf:li>`);
    parts.push('  </rdf:Seq>');
    parts.push('</dc:creator>');
  }

  if (info.subject) {
    parts.push('<dc:description>');
    parts.push('  <rdf:Alt>');
    parts.push(`    <rdf:li xml:lang="x-default">${escapeXml(info.subject)}</rdf:li>`);
    parts.push('  </rdf:Alt>');
    parts.push('</dc:description>');
  }

  if (info.keywords && info.keywords.length > 0) {
    parts.push('<dc:subject>');
    parts.push('  <rdf:Bag>');
    for (const kw of info.keywords) {
      parts.push(`    <rdf:li>${escapeXml(kw)}</rdf:li>`);
    }
    parts.push('  </rdf:Bag>');
    parts.push('</dc:subject>');
  }

  // XMP basic
  if (info.creationDate) {
    parts.push(`<xmp:CreateDate>${toISODate(info.creationDate)}</xmp:CreateDate>`);
  }
  if (info.modDate) {
    parts.push(`<xmp:ModifyDate>${toISODate(info.modDate)}</xmp:ModifyDate>`);
  }
  if (info.creator) {
    parts.push(`<xmp:CreatorTool>${escapeXml(info.creator)}</xmp:CreatorTool>`);
  }

  // PDF properties
  if (info.producer) {
    parts.push(`<pdf:Producer>${escapeXml(info.producer)}</pdf:Producer>`);
  }
  if (info.keywords && info.keywords.length > 0) {
    parts.push(`<pdf:Keywords>${escapeXml(info.keywords.join(', '))}</pdf:Keywords>`);
  }

  // PDF/A identification
  if (pdfALevel) {
    // Parse level like 'a', 'b', 'u' - the part number comes from the caller context
    parts.push(`<pdfaid:part>1</pdfaid:part>`);
    parts.push(`<pdfaid:conformance>${pdfALevel.toUpperCase()}</pdfaid:conformance>`);
  }

  parts.push('</rdf:Description>');
  parts.push('</rdf:RDF>');
  parts.push('</x:xmpmeta>');

  // Add padding for in-place editing
  for (let i = 0; i < 20; i++) {
    parts.push('                                                                                ');
  }

  parts.push('<?xpacket end="w"?>');

  return parts.join('\n');
}

/**
 * Embed XMP metadata as a stream object in the object store.
 * Returns a ref to the metadata stream.
 */
export function embedXMPMetadata(store: ObjectStore, xmp: string): PdfRef {
  const encoder = new TextEncoder();
  const data = encoder.encode(xmp);

  const ref = store.allocRef();
  const stream = pdfStream(
    {
      Type: pdfName('Metadata'),
      Subtype: pdfName('XML'),
    },
    data,
  );
  store.set(ref, stream);
  return ref;
}

File: src/metadata/xmp.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const DocumentInfo = @import("info_dict.zig").DocumentInfo;

/// Generate XMP (Extensible Metadata Platform) XML metadata from document info.
/// Returns a newly allocated XML string suitable for embedding as a PDF metadata stream.
pub fn generateXmp(allocator: Allocator, info: DocumentInfo) ![]u8 {
    var buf: std.ArrayListUnmanaged(u8) = .{};
    errdefer buf.deinit(allocator);
    const writer = buf.writer(allocator);

    try writer.writeAll("<?xpacket begin=\"\xEF\xBB\xBF\" id=\"W5M0MpCehiHzreSzNTczkc9d\"?>\n");
    try writer.writeAll("<x:xmpmeta xmlns:x=\"adobe:ns:meta/\">\n");
    try writer.writeAll("<rdf:RDF xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n");

    // Dublin Core properties
    try writer.writeAll("<rdf:Description rdf:about=\"\"\n");
    try writer.writeAll("  xmlns:dc=\"http://purl.org/dc/elements/1.1/\"\n");
    try writer.writeAll("  xmlns:xmp=\"http://ns.adobe.com/xap/1.0/\"\n");
    try writer.writeAll("  xmlns:pdf=\"http://ns.adobe.com/pdf/1.3/\"\n");
    try writer.writeAll("  xmlns:pdfaid=\"http://www.aiim.org/pdfa/ns/id/\">\n");

    if (info.title) |title| {
        try writer.writeAll("  <dc:title>\n");
        try writer.writeAll("    <rdf:Alt>\n");
        try writer.print("      <rdf:li xml:lang=\"x-default\">{s}</rdf:li>\n", .{title});
        try writer.writeAll("    </rdf:Alt>\n");
        try writer.writeAll("  </dc:title>\n");
    }

    if (info.author) |author| {
        try writer.writeAll("  <dc:creator>\n");
        try writer.writeAll("    <rdf:Seq>\n");
        try writer.print("      <rdf:li>{s}</rdf:li>\n", .{author});
        try writer.writeAll("    </rdf:Seq>\n");
        try writer.writeAll("  </dc:creator>\n");
    }

    if (info.subject) |subject| {
        try writer.writeAll("  <dc:description>\n");
        try writer.writeAll("    <rdf:Alt>\n");
        try writer.print("      <rdf:li xml:lang=\"x-default\">{s}</rdf:li>\n", .{subject});
        try writer.writeAll("    </rdf:Alt>\n");
        try writer.writeAll("  </dc:description>\n");
    }

    if (info.keywords) |keywords| {
        try writer.print("  <pdf:Keywords>{s}</pdf:Keywords>\n", .{keywords});
    }

    if (info.creator) |creator| {
        try writer.print("  <xmp:CreatorTool>{s}</xmp:CreatorTool>\n", .{creator});
    }

    if (info.producer) |producer| {
        try writer.print("  <pdf:Producer>{s}</pdf:Producer>\n", .{producer});
    }

    if (info.creation_date) |date| {
        try writer.print("  <xmp:CreateDate>{s}</xmp:CreateDate>\n", .{date});
    }

    if (info.mod_date) |date| {
        try writer.print("  <xmp:ModifyDate>{s}</xmp:ModifyDate>\n", .{date});
    }

    try writer.writeAll("</rdf:Description>\n");
    try writer.writeAll("</rdf:RDF>\n");
    try writer.writeAll("</x:xmpmeta>\n");
    try writer.writeAll("<?xpacket end=\"w\"?>");

    return buf.toOwnedSlice(allocator);
}

// -- Tests --

test "xmp: generate with all fields" {
    const allocator = std.testing.allocator;
    const xmp = try generateXmp(allocator, .{
        .title = "Test",
        .author = "Author",
        .subject = "Subject",
        .keywords = "test, pdf",
        .creator = "zpdf",
        .producer = "zpdf",
        .creation_date = "2024-01-01",
        .mod_date = "2024-01-02",
    });
    defer allocator.free(xmp);

    try std.testing.expect(std.mem.indexOf(u8, xmp, "<?xpacket begin=") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "<dc:title>") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "Test") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "<dc:creator>") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "<?xpacket end=") != null);
}

test "xmp: generate empty" {
    const allocator = std.testing.allocator;
    const xmp = try generateXmp(allocator, .{});
    defer allocator.free(xmp);

    try std.testing.expect(std.mem.indexOf(u8, xmp, "<?xpacket begin=") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "<dc:title>") == null);
}

test "xmp: generate partial" {
    const allocator = std.testing.allocator;
    const xmp = try generateXmp(allocator, .{ .title = "Only Title" });
    defer allocator.free(xmp);

    try std.testing.expect(std.mem.indexOf(u8, xmp, "Only Title") != null);
    try std.testing.expect(std.mem.indexOf(u8, xmp, "<dc:creator>") == null);
}
